### ISSUE TITLE
But ... Can RP2040 run Linux natively?

### ISSUE THREAD

**Issue by Sovenok-Hacker** [tvlad1234/pico-rv32ima#17]
If it have resources, why do community didn't port Linux on RP2040 natively?

**Comment by flyboy-byte** [comment]
for one the rp2040 dont have mmu. ud have to also port the kernal

**Comment by Sovenok-Hacker** [comment]
But ESP-32 doesn't have MMU, but Linux runs on it

**Comment by Sovenok-Hacker** [comment]
https://github.com/jcmvbkbc/linux-xtensa/issues/5

**Comment by flyboy-byte** [comment]
hm kernel stuff is above me. but i believe this project leverages a already written kernel. this linux xtensa on esp had to write a kernel to boot without mmu. correct me if im wrong byt u dont need a mmu if u write a kernel that dont readdress memory and the processes have direct access?

**Comment by flyboy-byte** [comment]
really cool project there. im reading through rn

**Comment by Mr-Bossman** [comment]
Chough shame full self promo https://github.com/Mr-Bossman/pi-pico2-linux

**Comment by tvlad1234** [comment]
> Chough shame full self promo https://github.com/Mr-Bossman/pi-pico2-linux
 Saw it today and I was just discussing it with my friends, that's amazing. Too bad the kernel requires MMU to run in XIP mode.

**Comment by Mr-Bossman** [comment]
> > Chough shame full self promo https://github.com/Mr-Bossman/pi-pico2-linux
> 
>  Saw it today and I was just discussing it with my friends, that's amazing.

Wait, u already saw my thing?

> Too bad the kernel requires MMU to run in XIP mode.

Yeah, I'm going to look into fixing that.

**Comment by tvlad1234** [comment]
Yes, somebody posted it in cnlohr's discord earlier today. We've been
discussing the no-MMU xip issue over the past week

lun., 26 aug. 2024, 23:06 Jesse Taube ***@***.***> a scris:

**Comment by Mr-Bossman** [comment]
> Yes, somebody posted it in cnlohr's discord earlier today.

Ah makes sence is there an invite link.
I hang here mostly https://discord.gg/embedded. or just email me

> We've been discussing the no-MMU xip issue over the past week 
Oh really?

**Comment by Mr-Bossman** [comment]
> I wanted to ask if it is possible/solved on the RP2040 chips to run linux natively without any virtualization?


As far as I know it not possible. It's not possible because you need like 4MB of RWX ram and the rp2040 doesn't let you do all three only RW or RX. Though iirc I think someone made a way to emulate writes by using the mpu to trap the writes  to the ram when. This is a valid way, but it emulates writes or just runs it again from internal sram.

**Comment by sivansh11** [comment]
@Mr-Bossman, I am pretty new to all this, I tried building the linux kernel with tinyconfig, thats around 700K, shouldnt that fit into rpi pico/w ?

**Comment by tvlad1234** [comment]
Closing, as discussion went off-topic.
